google-cloud-spanner/src/test/java/com/google/cloud/spanner/connection/DirectedReadTest.java

```diff
@@ -0,0 +1,251 @@
+                  IncludeReplicas.newBuilder()
+                      .addReplicaSelections(
+                          ReplicaSelection.newBuilder()
+                              .setLocation("eu-west1")
+                              .setType(ReplicaSelection.Type.READ_ONLY)
+                              .build())
+                      .build())
+              .build(),
+          DirectedReadOptions.newBuilder()
+              .setExcludeReplicas(
+                  ExcludeReplicas.newBuilder()
+                      .addReplicaSelections(
+                          ReplicaSelection.newBuilder()
+                              .setLocation("eu-west1")
+                              .setType(ReplicaSelection.Type.READ_ONLY)
+                              .build())
+                      .build())
+              .build(),
+          DirectedReadOptions.newBuilder().build()
+        }) {
+      try (Connection connection = createConnection()) {
+        connection.setAutocommit(true);
+        connection.execute(
+            Statement.of(
+                String.format(
+                    "set %sdirected_read='%s'",
+                    getVariablePrefix(), DirectedReadOptionsUtil.toString(expected))));
+        executeQuery(connection);
+        assertDirectedReadOptions(expected);
+
+        mockSpanner.clearRequests();
+
+        // Reset to default.
+        connection.execute(
+            Statement.of(String.format("set %sdirected_read=''", getVariablePrefix())));
+        executeQuery(connection);
+        assertDirectedReadOptions(DirectedReadOptions.getDefaultInstance());
+
+        mockSpanner.clearRequests();
+      }
+    }
+  }
+
+  @Test
+  public void testDirectedReadIsIgnoredInReadWriteTransaction() {
+    try (Connection connection = createConnection()) {
+      connection.setAutocommit(false);
+      connection.setReadOnly(false);
+
+      connection.execute(
+          Statement.of(
+              String.format(
+                  "set %sdirected_read='%s'",
+                  getVariablePrefix(),
+                  DirectedReadOptionsUtil.toString(
+                      DirectedReadOptions.newBuilder()
+                          .setIncludeReplicas(
+                              IncludeReplicas.newBuilder()
+                                  .addReplicaSelections(
+                                      ReplicaSelection.newBuilder()
+                                          .setType(ReplicaSelection.Type.READ_WRITE)
+                                          .setLocation("us-west1")
+                                          .build())
+                                  .build())
+                          .build()))));
+      // This uses a read/write transaction, which will ignore any DirectedReadOptions.
+      executeQuery(connection);
+      assertDirectedReadOptions(DirectedReadOptions.getDefaultInstance());
+    }
+  }
+
+  @Test
+  public void testDirectedReadIsUsedInReadOnlyTransaction() {
```

Comment:
Can we also test that DirectedReadOptions is persisted across different statements/transactions in a connection once set until reset?

Reply:
Good point, added both for read-only transactions and auto-commit.